### PR TITLE
Messenger support

### PR DESCRIPTION
**Pull request by mynameisbogdan**
| Q | A
| --- | ---
| Branch? | 2.x 
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes
| Fixed tickets | #1193 
| License | MIT
| Doc PR | 

Adding optional `symfony/messenger` support.

### CODE REVIEW

**Comment by peter-gribanov** [comment]
Perhaps you can use the [EnqueueBundle](https://symfony.com/doc/2.x/bundles/LiipImagineBundle/resolve-cache-images-in-background.html) documentation as a basis to document this feature.

**Comment by alex-cool-tech** [comment]
Hi, any updates maybe?

**Comment by comxd** [comment]
I have a problem using flysystem and Google Cloud Storage:

```
"error" => "Handling "App\Message\LiipImageResolveCacheMessage" failed: The mime type of file path/to/my/image.jpg?itok=yeDn2W0I must be image/xxx or application/pdf, got application/octet-stream."
```

**Review by dbu — APPROVED** [review]
thanks a lot, this looks good!

can you please add documentation (duplicate and adjust what we have for enqueue-bundle)?

then we can merge this, and we will deprecate the enqueue integration.

File: Message/Handler/ResolveCacheHandler.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public function __invoke(ResolveCache $message): void
+    {
+        $filters = $message->getFilters() ?: array_keys($this->filterManager->getFilterConfiguration()->all());
+        $path = $message->getPath();
+
+        foreach ($filters as $filter) {
+            if ($message->isForce()) {
+                $this->filterService->bustCache($path, $filter);
```
Comment (dbu) [inline Message/Handler/ResolveCacheHandler.php]: likely needs to be adjusted to the change in https://github.com/liip/LiipImagineBundle/pull/1347 once that is merged.

Comment (mynameisbogdan) [reply Message/Handler/ResolveCacheHandler.php]: Ok, I'll change it then.

**Comment by mynameisbogdan** [comment]
@dbu done regarding the docs.

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 2]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
   The varnish may cache the page with those links to the resolve controller.
   A browser keeps sending requests to it though there is no need for it after the first call.
 
+Messenger
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
Symfony Messenger
```

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 3]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
...
+-------------------
+
+Step 1: Install
+~~~~~~~~~~~~~~~~
+
+First, we have to `install symfony/messenger`_. You have to basically use composer to install the Symfony component.
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
First, `install symfony/messenger`_ with composer:
```

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 4]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
...
+                # https://symfony.com/doc/current/messenger.html#transport-configuration
+                async: '%env(MESSENGER_TRANSPORT_DSN)%'
+                sync: 'sync://'
+
+            routing:
+                # Route your messages to the transports
+                'Liip\ImagineBundle\Message\ResolveCache': async
+
+Step 2: Configure LiipImagineBundle
+~~~~~~~~~~~~~~~~
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]: the underlines should be the same length as the title

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 5]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
...
+
+    # use -vv to see details about what's happening
+    $ php bin/console messenger:consume async --time-limit=3600 --memory-limit=256M -vv
+
+Step 4: Send resolve cache message
+~~~~~~~~~~~~~~~~
+
+You have to dispatch a message in order to process images in background.
+The message must contain the original image path (in terms of LiipImagineBundle).
+If you do not define filters the background process will resolve cache for all available filters.
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
If you do not define filters, the background process will resolve cache for all available filters.
```

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 6]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
...
+If the cache already exist the background process does recreate it by default
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
If the cache already exists, the background process does recreate it by default
```

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 7]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
...
+You can force cache to be recreated and in this case the cached image is removed and a new one replaces it.
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst:77]: is the above line incorrect then? by default, we do *not* recreate the image, you need to force to recreate?

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 8]
```diff
@@ -17,10 +14,98 @@ Though there are some disadvantages:
   The varnish may cache the page with those links to the resolve controller.
   A browser keeps sending requests to it though there is no need for it after the first call.
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst:17]: i think we should explain a bit more how the solution works here:

```suggestion
To prepare the cached images in advance, the LiipImagineBundle allows you to use a message queue to have a worker warm up the cache asynchronously. Your application has to send messages about the images as it becomes aware of them (file upload, import processes, ...) and you need to run the worker for the message queue.
```

File: Message/Handler/ResolveCacheHandler.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+namespace Liip\ImagineBundle\Message\Handler;
+
+use Liip\ImagineBundle\Imagine\Filter\FilterManager;
+use Liip\ImagineBundle\Message\ResolveCache;
+use Liip\ImagineBundle\Service\FilterService;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+/** @experimental */
```
Comment (dbu) [inline Message/Handler/ResolveCacheHandler.php]:
```suggestion
/** 
 * Listen to ResolveCache messages and prepare the cache for those images.
 *
 * @experimental
 */
```

File: Message/ResolveCache.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * (c) https://github.com/liip/LiipImagineBundle/graphs/contributors
+ *
+ * For the full copyright and license information, please view the LICENSE.md
+ * file that was distributed with this source code.
+ */
+
+namespace Liip\ImagineBundle\Message;
+
+/** @experimental */
```
Comment (dbu) [inline Message/ResolveCache.php]:
```suggestion
/**
 * Message to warm up the cache for an image.
 * 
 * Your application needs to dispatch this message when it becomes aware of a new image.
 *
 * @experimental 
 */
```

**Review by dbu — APPROVED** [review]
thanks a lot, this looks good!

i have some suggestions, and please make the underlines match the length of the titles in the .rst doc.

i am a bit unhappy with the naming of "ResolveCache" when what we really do is warm up the cache. resolve is the word in the loaders, and it has the side effect of filling the cache. the unelegant terminology is already used with enqueue, but i intend to deprecate that enqueue integration once your PR is merged. what would you think of renaming the message to WarmupCache and the handler to CacheWarmupHandler?

File: Resources/doc/resolve-cache-images-in-background.rst (hunk 7 above)
Comment (mynameisbogdan) [reply Resources/doc/resolve-cache-images-in-background.rst:77]: Yes, seems incorrect. We do not recreate the image, you have set force to true for that.

**Comment by mynameisbogdan** [comment]
Great @dbu! 

I implemented your suggestions with one little change. I renamed the message to `WarmupCache`, and the handler to `WarmupCacheHandler` to stick to Symfony's naming convention of `${messageClassName}Handler`. 

Also in the docs, I changed some `resolve` wording to `warmup`. And recommended a new transport named `liip_imagine`.

It's this okay with you?

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 11]
```diff
@@ -16,11 +13,100 @@ Though there are some disadvantages:
   If there is nothing in the cache the page will contain the url to resolve controller.
   The varnish may cache the page with those links to the resolve controller.
   A browser keeps sending requests to it though there is no need for it after the first call.
+To prepare the cached images in advance, the LiipImagineBundle allows you to use a message queue to have a worker warm up the cache asynchronously. Your application has to send messages about the images as it becomes aware of them (file upload, import processes, ...) and you need to run the worker for the message queue.
+
+Symfony Messenger
+-------------------
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
-----------------
```

File: Resources/doc/resolve-cache-images-in-background.rst [hunk 12]
```diff
@@ -16,11 +13,100 @@ Though there are some disadvantages:
...
+
+.. code-block:: terminal
+
+    $ php bin/console messenger:consume liip_imagine --time-limit=3600 --memory-limit=256M
+
+    # use -vv to see details about what's happening
+    $ php bin/console messenger:consume liip_imagine --time-limit=3600 --memory-limit=256M -vv
+
+Step 4: Send warmup cache message
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (dbu) [inline Resources/doc/resolve-cache-images-in-background.rst]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Review by dbu — APPROVED** [review]
thanks a lot!

**Pull request merged**